test/unit/index/fm_index_iterator_test.cpp

```diff
@@ -0,0 +1,343 @@
+    EXPECT_EQ(it.query_length(), 1);
+
+    EXPECT_TRUE(it.extend_right());
+    EXPECT_TRUE(is_set_equal(it.locate(), (std::vector<uint64_t>{1, 4})));
+    EXPECT_EQ(it.query_length(), 2);
+
+    // unsuccessful cycle_back(), it remains untouched
+    TypeParam it_cpy = it;
+    EXPECT_FALSE(it.cycle_back());
+    EXPECT_EQ(it, it_cpy);
+
+    // unsuccessful extend_right(), it remains untouched
+    it = TypeParam(fm);
+    EXPECT_TRUE(it.extend_right("GACG"_dna4));
+    it_cpy = it;
+    EXPECT_FALSE(it.extend_right());
+    EXPECT_EQ(it, it_cpy);
+
+    // cycle_back() cannot be called on the root node
+    it = TypeParam(fm);
+    #ifndef NDEBUG
+        EXPECT_DEATH(it.cycle_back(), "");
+    #endif
+    EXPECT_EQ(it, TypeParam(fm));
+}
+
+TYPED_TEST(fm_index_iterator_test, query)
+{
+    typename TypeParam::index_type::text_type text{"ACGACG"_dna4};
+    typename TypeParam::index_type fm{text};
+
+    // query()
+    TypeParam it(fm);
+    EXPECT_TRUE(it.extend_right("ACG"_dna4));
+    EXPECT_TRUE(ranges::equal(*it, "ACG"_dna4));
+    EXPECT_TRUE(ranges::equal(it.query(), "ACG"_dna4));
+}
+// TODO: test last_char()
+TYPED_TEST(fm_index_iterator_test, incomplete_alphabet)
+{
+    // search a char that does not occur in the text (higher rank than largest char occurring in text)
+    {
+        typename TypeParam::index_type::text_type text{"ACGACG"_dna4};
+        typename TypeParam::index_type fm{text};
+        TypeParam it = TypeParam(fm);
+        EXPECT_FALSE(it.extend_right(dna4::T));
+        EXPECT_EQ(it, TypeParam(fm));
+    }
+
+    // search a char that does not occur in the text (smaller rank than smallest char occurring in text)
+    {
+        typename TypeParam::index_type::text_type text{"CGTCGT"_dna4};
+        typename TypeParam::index_type fm{text};
+        TypeParam it = TypeParam(fm);
+        EXPECT_FALSE(it.extend_right(dna4::A));
+        EXPECT_EQ(it, TypeParam(fm));
+    }
+
+    // search a char that does not occur in the text
+    // (some rank that is neither the smallest nor the highest occurring in text)
+    {
+        typename TypeParam::index_type::text_type text{"ATATAT"_dna4};
+        typename TypeParam::index_type fm{text};
+        TypeParam it = TypeParam(fm);
+        EXPECT_FALSE(it.extend_right(dna4::C));
+        EXPECT_FALSE(it.extend_right(dna4::G));
+        EXPECT_FALSE(it.extend_right("ACGT"_dna4));
+        EXPECT_FALSE(it.extend_right("G"_dna4));
+        EXPECT_EQ(it, TypeParam(fm));
+
+        EXPECT_TRUE(it.extend_right(dna4::A));
+        EXPECT_TRUE(it.cycle_back());
+        EXPECT_TRUE(ranges::equal(it.query(), "T"_dna4));
+    }
+}
+
+TYPED_TEST(fm_index_iterator_test, lazy_locate)
+{
+    typename TypeParam::index_type::text_type text{"ACGTACGT"_dna4};
+    typename TypeParam::index_type fm{text};
+
+    TypeParam it = TypeParam(fm);
+    it.extend_right("ACG"_dna4);
+
+    EXPECT_TRUE(ranges::equal(it.locate(), it.lazy_locate()));
+}
+
+// TODO: tests for other alphabets
```

Comment:
can you do a typed test with different alphabet?

Reply:
I don't see a generic way to do this except come up with new sequences and queries for a different alphabet. But since we also have a test for incomplete alphabets, the benefit is rather small.